polkadot/node/network/collator-protocol/src/validator_side/mod.rs

```diff
@@ -2110,6 +2155,223 @@ async fn handle_collation_fetch_response(
+	// block hash including the block hash itself.
+	state
+		.implicit_view
+		.known_allowed_relay_parents_under(relay_parent, Some(*para_id))
+		.map(|ancestors| {
+			ancestors
+				.iter()
+				.take(claim_queue_len)
+				.map(|anc| state.seconded_and_pending_for_para(anc, para_id))
+				.sum()
+		})
+		.unwrap_or(0)
+}
+
+// Iterates all relay parents under `relay_parent` within the view up to `claim_queue_len` and
+// returns the number of claims for `para_id` which are affecting `relay_parent`. Affecting in this
+// context means that a relay parent before `relay_parent` have claimed a slot from `relay_parent`'s
+// view.
+fn claimed_within_view(
+	state: &State,
+	relay_parent: &Hash,
+	para_id: &ParaId,
+	claim_queue_len: usize,
+) -> usize {
+	// First we take all the ancestors of `relay_parent`. The order is `relay_parent` -> ancestor...
+	let ancestors = if let Some(anc) = state
+		.implicit_view
+		.known_allowed_relay_parents_under(relay_parent, Some(*para_id))
+	{
+		anc
+	} else {
+		return 0;
+	};
+
+	// We want to take the ancestors which fall in the claim queue view (hence the `take()`) and
+	// start counting claims from the oldest relay parent (hence the `rev()`). We start from oldest
+	// so that we can track claims from the ancestors.
+	// Example:
+	// CQ: [A, B, A]
+	// Relay parents: 1, 2, 3.
+	// We want to second a candidate for para A at relay parent 3.
+	// At relay parent 1 we have got two advertisements for A and one for B.
+	// At relay parent 2 we have got no advertisements.
+	// At relay parent 3 we CHECK if we can accept an advertisement for A.
+	// Visually:
+	//
+	// CQ: [A B A]    CQ: [B A B]    CQ: [A B A]
+	//
+	// ┌───────┐      ┌───────┐      ┌───────┐
+	// │  RP1  ┼──────►  RP2  ┼──────►  RP3  │
+	// └───────┘      └───────┘      └───────┘
+	//   A1
+	//   B1────────────►
+	//   A2───────────────────────────►
+	//
+	// Since we have got three advertisements at RP1, A1 claims the slot at RP1, B1 - at RP2 and A2
+	// - at RP3. This means that at RP2 we still can accept one more advertisement for B and one for
+	// A.
+	// How the counting must work for para id A. ACC here represents the number of claims
+	// transferred to the next relay parent which finally are transferred to the target relay
+	// parent:
+	// - at RP1 we have got two claims and para A is in the first position in the claim queue. ACC =
+	//   2 - 1 = 1.
+	// - at RP2 we have got no claims advertised. Para A is not on the first position of the claim
+	//   queue, so A2 can't be scheduled here. ACC is still 1.
+	// - at RP 3 we reach our target relay parent with ACC = 1. We want to advertise at RP 3 so we
+	//   add 1 to ACC and it becomes 2. IMPORTANT: since RP3 is our target relay parent we DON'T
+	//   subtract 1 since this is the position the new advertisement will occupy.
+	// - ACC = 2 is the final result which represents the number of claims for para A at RP3.
```

Comment:
as discussed, the comment needs updating

Reply:
Fixed